Make Get-ChildItem follow symlinks on demand, with checks for link loops

jeffbi opened the pull request:
Fixes #3951

    Add the dynamic parameter -FollowSymlink to Get-ChildItem.
    Add a mechanism for tracking visited directories.
    Add native code to get device/inode information on Unix/Windows.
    Add warning when refusing to enter an already-visited directory.

Commit:
Make Get-ChildItem follow symlinks on demand, with checks for link loops. (#3951)
    Add -FollowSymlink switch parameter to Get-ChildItem.
    Add a mechanism for tracking visited directories.
    Add native code to get dev/inode information on Unix.
    Add warning when refusing to enter an already-visited directory.

jeffbi commented:
@iSazonov, @daxian-dbw can someone look at the AppVeyor failure? Looking at the details, It appears to be failing _very_ early in the run

iSazonov commented:
@jeffbi I believe it is temporary error - please restart CI by reward your commit.

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -8238,6 +8330,47 @@ internal static bool WinIsSameFileSystemItem(string pathOne, string pathTwo)
+                {
+                    BY_HANDLE_FILE_INFORMATION info;
+
+                    if (GetFileInformationByHandle(sf.DangerousGetHandle(), out info))
```
Is it supported on Windows 7?

jeffbi replied:
It's supported as early as XP.

iSazonov replied:
Closed.

jeffbi commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1539,8 +1605,15 @@ protected override void InvokeDefaultAction(string path)
                     DirectoryInfo directory = new DirectoryInfo(path);
+                    InodeTracker tracker = null;
+
+                    if (recurse && Context.MyInvocation.BoundParameters.ContainsKey("FollowSymlink"))
```
There are two parameter sets for `Get-ChildItem`: `Items` and `LiteralItems`. The `-Recurse` and `-FollowSymlink` parameters are in both parameter sets.

The `Depth` parameter, which is also in both, implies `-Recurse` by setting the `Recurse` property. Should `-FollowSymlink` follow that pattern? That would remove the check of `recurse` here.

iSazonov replied:
We can not define explicitly mandatory sub-parameters 😕 That would be a useful feature.
So it is more safe to leave "recurse" here.

If `-Depth`  implies  `-Recurse` I believe we can do the same with `-FollowSymlink `.

jeffbi replied:
Modified the `-FollowSymlink` parameter to set `-Recurse`, and the `-Depth` parameter does.

iSazonov replied:
> Context.MyInvocation.BoundParameters.ContainsKey("FollowSymlink")) 

Can we use `FollowSymlink` property?

jeffbi replied:
FollowSymlink is a property of `GetChildItemCommand`, not of the `FileSystemProvider`.

iSazonov replied:
Thanks for clarify!

Closed.

jeffbi commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1608,9 +1681,13 @@ protected override void InvokeDefaultAction(string path)
             ReturnContainers returnContainers)
         {
+            if (tracker != null)
+                tracker.Visit(directory.FullName);
```
This function is called recursively, so I think we need to leave these lines here.

iSazonov replied:
`tracker = new InodeTracker();` - It's not expensive. We can always create this on top level and not check many times during recursion.

jeffbi replied:
In the `Dir` function, the `tracker` object also acts as a flag that the user gave the `-FollowSymlink` parameter.

iSazonov replied:
I tried to refactor code and had no benefits.

Closed.

jeffbi commented on src/libpsl-native/src/getinodedata.cpp:
```diff
@@ -0,0 +1,56 @@
+#include <unistd.h>
+#include <string>
+
+//! @brief GetInodeData retrieves a file's device and inode information.
```
This one describes the function. The one at the top of the file describes the file.

I used the existing `getlinkcount.cpp` as the pattern for this.

iSazonov replied:
Closed.

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1788,12 +1865,27 @@ protected override void InvokeDefaultAction(string path)
+                            bool enterDir = true;
+                            if (tracker != null)
+                            {
+                                if (tracker.Visited(recursiveDirectory.FullName))
```
It seems we can exclude ` enterDir`:

```c#
                            if (tracker != null)
                            {
                                if (tracker.Visited(recursiveDirectory.FullName))
                                {
                                    WriteWarning(StringUtil.Format(FileSystemProviderStrings.AlreadyListedDirectory,
                                                                    recursiveDirectory.FullName));
                                    continue;
                                }
                            }
                            else
                            {
                                if (InternalSymbolicLinkLinkCodeMethods.IsReparsePoint(recursiveDirectory))
                                {
                                    continue;
                                }
                            }

                            bool hidden = false;
                            if (!Force)
                            {
                                hidden = (recursiveDirectory.Attributes & FileAttributes.Hidden) != 0;
                            }

                            // if "Hidden" is explicitly specified anywhere in the attribute filter, then override
                            // default hidden attribute filter.
                            if (Force || !hidden || isFilterHiddenSpecified || isSwitchFilterHiddenSpecified)
                            {
                                Dir(recursiveDirectory, recurse, depth - 1, tracker, nameOnly, returnContainers);
                            }
```

jeffbi replied:
Refactored code to eliminate `enterDir`.

iSazonov replied:
Closed.

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1788,12 +1865,27 @@ protected override void InvokeDefaultAction(string path)
+                            {
+                                if (tracker == null)
+                                {
+                                    enterDir = false;
```
Closed.

jeffbi commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -31,6 +31,72 @@
+            {
+                if (_visitations.ContainsKey(dev))
+                {
+                    return _visitations[dev].ContainsKey(inode);
```
Part of that discussion ([https://stackoverflow.com/questions/2877660/composite-key-dictionary#comment14110652_2877820](https://stackoverflow.com/questions/2877660/composite-key-dictionary#comment14110652_2877820)) suggests that the nested dictionary approach may be faster.

As for the C#7 pattern, I thought I read somewhere in another code review that C#7 code should be avoided. Is that not the case?

iSazonov replied:
>C#7 pattern

We have no strict documentation. Different maintainers may have different requirements. My understanding: New patterns are welcome for the new code. Existing patterns (not relevant to the PR) should not be changed without maintainer's requiring.

> the nested dictionary approach may be faster.

There, the main answer has recently been updated and there is a reference to [gist with the tests](https://gist.github.com/ljw1004/61bc96700d0b03c17cf83dbb51437a69). The conclusion is now Tuple keys is fast.
We both think that performance is important here. So maybe it makes sense to spend time testing cases.
Another thought is, in most cases, the user will have one volume. We could use one dictionary until we find a reference to another volume. Although it will complicate the code and it's not clear how much we win. Ooops, for other volume we can make separate recursive `Dir()` call - in the case we have to switch Dictionary (tracker) only on "top" level when we meet new volume. In other words, we will have two dictionaries but they will not be nested and the number of comparisons will be less - we'll be faster.

And maybe rename `Dir()` in something like `DirectoryVisitor()` ?

jeffbi replied:
It may be that on Windows the user will have only one volume, but that is much less the case on Unix. For example, on my Linux box the root (`/`) directory has device ID 2049, the `/dev` directory has ID 6, and the `/proc` directory has ID 4. Symlinks can cross device/volume boundaries.

I've changed the code to use the C#7 ValueTuple mechanism.

Commit:
Changes per code review:

  o Switched from nested dictionary to single dictionary keyed by
    ValueTuple (device, inode).
  o Modified warning message to put path at end.
  o -FollowSymlink switch now implies -Recurse
  o Other code and comment refactorizings.

jeffbi commented on src/Microsoft.PowerShell.Commands.Management/commands/management/GetChildrenCommand.cs:
```diff
@@ -172,6 +172,23 @@ public uint Depth
+            }
+            set
+            {
+                _followSymlink = value;
```
Modified the `-FollowSymlink` parameter to set `-Recurse`, and the `-Depth` parameter does. Thus we cannot use auto-property here.

iSazonov replied:
Closed.

jeffbi commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1788,12 +1865,27 @@ protected override void InvokeDefaultAction(string path)
+                            }
+                            // We only want to recurse into symlinks if a) the user has asked to with
+                            // the -FollowSymLinks switch parameter and b) the directory the symlink
+                            // points to has not already been visited, preventing symlink loops.
```
Modified the comment.

iSazonov replied:
Closed.

jeffbi commented on src/System.Management.Automation/resources/FileSystemProviderStrings.resx:
```diff
@@ -339,4 +339,7 @@
     <value>Cannot create symbolic link because the path {0} already exists.</value>
   </data>
+  <data name="AlreadyListedDirectory" xml:space="preserve">
+    <value>Directory {0} has already been listed.</value>
```
Re-formatted the message to put the path at the end.

iSazonov replied:
Closed.

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -8238,6 +8321,47 @@ internal static bool WinIsSameFileSystemItem(string pathOne, string pathTwo)
             return false;
         }
 
+        internal static bool GetInodeData(string path, out UInt64 dev, out UInt64 inode)
```
Now we don't use dev and inode and we could out Tuple.

jeffbi replied:
We do still use dev and inode. That's how we uniquely identify a file system object.

iSazonov replied:
I see the `GetInodeData` is only called in `Visit` and `Visited` method.

jeffbi replied:
Yes. Did you want to move the implementation up into the `InodeTracker` class?

iSazonov replied:
I meant simple change like ``` internal static bool GetInodeData(string path, out (dev, inode) var) ```

iSazonov replied:
Closed.

iSazonov commented on src/System.Management.Automation/resources/FileSystemProviderStrings.resx:
```diff
@@ -339,4 +339,7 @@
     <value>Cannot create symbolic link because the path {0} already exists.</value>
   </data>
+  <data name="AlreadyListedDirectory" xml:space="preserve">
+    <value>Not listing already-listed directory {0}.</value>
```
Can we exclude repetition  "listing ... listed"?

jeffbi replied:
Changed "already-listed" to "already-visited"

iSazonov replied:
Closed.

Commit:
More changes per code review

jeffbi commented on src/Microsoft.PowerShell.Commands.Management/commands/management/GetChildrenCommand.cs:
```diff
@@ -172,6 +172,25 @@ public uint Depth
+            {
+                _followSymlink = value;
+                if (value == true)
+                    Recurse = true;
```
It looks like that was in response to a specific bug. In our case it's just part of a new feature.

iSazonov replied:
I'm afraid we could catch the same bug here. Maybe maintainers clarify.

iSazonov commented:
@daxian-dbw Could you please assign a reviewer?

Commit:
More changes per code review

jeffbi commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -31,6 +31,61 @@
+        public bool Visit(string path)
+        {
+            var inodeData = (0UL, 0UL);
+            if (InternalSymbolicLinkLinkCodeMethods.GetInodeData(path, out inodeData))
```
Thanks for catching that. New line inserted.

iSazonov replied:
Closed.

jeffbi commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -8238,6 +8317,46 @@ internal static bool WinIsSameFileSystemItem(string pathOne, string pathTwo)
+                    {
+                        UInt64 tmp = info.FileIndexHigh;
+                        tmp <<= 32;
+                        tmp |= info.FileIndexLow;
```
Made the calculation more concise.

iSazonov replied:
Closed.

iSazonov commented:
@jeffbi Great work! Thanks!
I leave two questions opened for maintainers.

LGTM.

daxian-dbw commented:
@jeffbi and @iSazonov thanks for the good work and review! I will quickly go through the changes before merging.

daxian-dbw commented on src/System.Management.Automation/resources/FileSystemProviderStrings.resx:
```diff
@@ -339,4 +339,7 @@
     <value>Cannot create symbolic link because the path {0} already exists.</value>
   </data>
+  <data name="AlreadyListedDirectory" xml:space="preserve">
+    <value>Not listing already-visited directory {0}.</value>
```
I think `Skip already-visited directory {0}` might be better.

jeffbi replied:
Should this be "Skip"? Or perhaps "Skipping" or "Skipped"?

jeffbi replied:
Fixed.

daxian-dbw commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -8238,6 +8318,45 @@ internal static bool WinIsSameFileSystemItem(string pathOne, string pathTwo)
         }
 
+        internal static bool GetInodeData(string path, out System.ValueTuple<UInt64, UInt64> inodeData)
+        {
```
I see that you are following a pattern like `IsSameFileSystemItem` or `IsHardLink`. Ideally, code here should be refactored to consolidate Unix/Windows implementations in one place and eliminate wrapper functions like `IsSameFileSystemItem` and `IsHardLink`. Instead they should be
```CSharp
internal static bool IsSameFileSystemItem(string pathOne, string pathTwo)
{
#if UNIX
   // Unix implementation goes here
#else
   // Windows implementation goes here
#endif
}
```
I'm not asking you to do this change for this PR, just want to call out that it should be this way moving forward. I opened #4086 to track the refactoring task.

daxian-dbw commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -31,6 +31,62 @@
+        /// </summary>
+        /// <param name="path">Path to the file or directory to be marked as visited.</param>
+        /// <returns>True if the path was successfully added to the visited list, false otherwise</returns>
+        public bool Visit(string path)
```
Correspondingly, maybe this name can be changed to `VisitPath`.

daxian-dbw replied:
And, does this method really need to return anything? If `GetInodeData` failed, then we just don't skip anything.

jeffbi replied:
Fixed both.

jeffbi commented on src/Microsoft.PowerShell.Commands.Management/commands/management/GetChildrenCommand.cs:
```diff
@@ -172,6 +172,25 @@ public uint Depth
+        /// Gets or sets whether symbolic links are to be followed when recursing.
+        /// </summary>
+        [Parameter]
+        public SwitchParameter FollowSymlink
```
So then `-FollowSymlink` should not imply `-Recurse`? It has meaning only if `-Recurse` is also explicitly given?

iSazonov replied:
powercode initiated #4102 for disscusion of parameter sets selection. So it is design case whether or not imply "mandatory" parameter by subparameter.

jeffbi replied:
Event without `-FollowSymlink`, the cmdlet will follow symlinks specified in `-Path`

iSazonov replied:
Should we follow w/o FollowSymlink ?

daxian-dbw replied:
Maybe it will turn out that making `-FollowSymlink` implicitly imply `-Recurse` is better, but for now here are my 2 cents:
1. It's always better to force the user to be explicit about their intent, so `dir -Recurse -FollowSymlink` is more readable than `dir -FollowSymlink` in my opinion.
2. For simplicity reason, this PR can just add the `-FollowSymlink` parameter without worrying about its dependency to `-Recurse`. In my understanding (correct me if I'm wrong) the related code won't execute when it's not recursively iterating file system items. Then we can get feedback from users and base on that to decide if we should make `-FollowSymlink` implicitly imply `-Recurse`.

jeffbi replied:
Changed to dynamic parameter. `-FollowSymlink` no longer implies `-Recurse`.

Commit:
Changes per code review.

jeffbi commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -31,6 +31,62 @@
+        /// <summary>
+        /// Construct a new InodeTracker
+        /// </summary>
+        public InodeTracker()
```
Fixed.

jeffbi commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -31,6 +31,62 @@
+        /// </summary>
+        /// <param name="path">Path to the file or directory to be checked.</param>
+        /// <returns>True if the path has been visited, false otherwise</returns>
+        public bool Visited(string path)
```
Fixed.

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -68,19 +68,13 @@ public bool Visited(string path)
         /// </summary>
         /// <param name="path">Path to the file or directory to be marked as visited.</param>
         /// <returns>True if the path was successfully added to the visited list, false otherwise</returns>
```
Now we return nothing. Please remove the line.

iSazonov replied:
Closed.

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1676,7 +1674,7 @@ protected override void InvokeDefaultAction(string path)
         {
             if (tracker != null)
-                tracker.Visit(directory.FullName);
+                tracker.VisitPath(directory.FullName);
```
Wrong pattern.
We can use `tracker?.VisitPath(directory.FullName); ` or `If (...) { ... }`.

iSazonov replied:
Closed.

Commit:
More changes per code review.

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1788,26 +1854,39 @@ protected override void InvokeDefaultAction(string path)
+                                    WriteWarning(StringUtil.Format(FileSystemProviderStrings.AlreadyListedDirectory,
+                                                                recursiveDirectory.FullName));
+                                    continue;
                                 }
```
Sorry, I again under 'tracker == null' doulble check pressure.
It seem we can remove one if remove lines 1675-1678, add its before first `Dir` call and here add:
```c#
                                else
                                {
                                    tracker.VisitPath(directory.FullName);
                                }
```

Next, we can exclude `VisitPath` at all if make "visit" in `IsPathVisited` (we should rename it maybe as `TryPathVisit`).

jeffbi replied:
I don't agree with this. If we mark the directory has having been visited here, then if the test at line 1883 evaluates to false the `Dir` method will not be called and we will not visit the directory even though we have already claimed to have done so.

iSazonov replied:
Do you mean "hidden" directories? If so we can safely mark "hidden" directories as "visited" - in both cases they is really skipped.

jeffbi replied:
A directory might be hidden. A symbolic link (or several such) might not be.

iSazonov replied:
You are right. We can move the hidden check above line 1861 - it is less expensive then double call 'GetInodeData'.

Commit:
Additional changes per code review.

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -40,41 +40,49 @@ internal class InodeTracker
             if (InternalSymbolicLinkLinkCodeMethods.GetInodeData(path, out inodeData))
             {
-                _visitations[inodeData] = true;
+                if (_visitations.ContainsKey(inodeData))
```
Now CoreFX has a new method [Dictionary.TryAdd](https://github.com/dotnet/corefx/blob/3d4d8038d1a3f0c56d3046bc7494f30ecc35e67b/src/Common/tests/System/Collections/DictionaryExtensions.cs) - Could we use them?

jeffbi replied:
Used new method

Commit:
Use Dictionary's TryAdd extension method.

iSazonov commented:
LGTM.

@daxian-dbw Could you please continue the code review?

daxian-dbw commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -31,6 +31,56 @@
+        /// If the path has not yet been visited then before this method returns it marks
+        /// the path as having been visited.
+        /// </remarks>
+        internal bool TryVisitPath(string path)
```
This method goes against the convention of `Try<Verb>` pattern. Take `TryAdd` as an instance, if added, then return true; if not added, then return false. Following that conversion, `TryVisitPath` should be: if visited, then return true; if not visited, then return false. So `TryVisitPath` makes it confusing understanding the code.

How about making it `IsPathVisited(bool visitIfNotAlready = false)`?
```CSharp
internal bool IsPathVisited(bool visitIfNotAlready = false)
{
    bool isPathVisited = false;
    if (InternalSymbolicLinkLinkCodeMethods.GetInodeData(path, out (UInt64, UInt64) inodeData))
    {
        if (visitIfNotAlready)
        {
            isPathVisited = !_visitations.Add(inodeData)
        }
        else
        {
            isPathVisited = _visitations.Contains(inodeData)
        }
    }
    return isPathVisited;
}
```

iSazonov replied:
With HashSet it will looks just better.

jeffbi replied:
Fixed.

daxian-dbw commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -31,6 +31,56 @@
+    /// <summary>
+    /// Tracks visited files/directories by caching their device IDs and inodes.
+    /// </summary>
+    internal class InodeTracker
```
This class seems to be a helper that will only be used in FileSystemProvider code, so can we make it a private nested class within FileSystemProvider? (just like `private static class NativeMethods` in FileSystemProvider)

jeffbi replied:
Fixed.

daxian-dbw commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -31,6 +31,56 @@
+            _visitations = new Dictionary<(UInt64, UInt64), bool>();
+            var inodeData = (0UL, 0UL);
+
+            if (InternalSymbolicLinkLinkCodeMethods.GetInodeData(path, out inodeData))
```
Minor comment: maybe use the following pattern to avoid `var inodeData = (0UL, 0UL);`?
```CSharp
if (InternalSymbolicLinkLinkCodeMethods.GetInodeData(path, out (UInt64, UInt64) inodeData))
```

iSazonov replied:
Looks good.

jeffbi replied:
Fixed.

daxian-dbw reviewed (commented):
Thanks for switching to the dynamic parameter approach! See some more comments below.

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -31,6 +31,56 @@
+    /// </summary>
+    internal class InodeTracker
+    {
+        private Dictionary<(UInt64, UInt64), bool>  _visitations;
```
We use new Dictionary.TryAdd() but HashSet.Add() works the same - so we can use HashSet.

jeffbi replied:
HashSet is much better. Fixed.

Commit:
Several changes per code review
   * Make InodeTracker a private nested class
   * Change from Dictionary to HashSet in InodeTracker
   * Change InodeTracker's TryVisitPath to IsPathVisited with a parameter for marking a path as visited if not already
   * Add check for -Recurse before creating InodeTracker object
   * Change order of parameters in Dir method
   * Remove unneeded nested block level
   * Add missing </summary> tag
   * Add -Pending on some tests to get past Pester errors
       (see issue #4145)

jeffbi commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1539,8 +1589,16 @@ protected override void InvokeDefaultAction(string path)
                     DirectoryInfo directory = new DirectoryInfo(path);
+                    InodeTracker tracker = null;
+
+                    GetChildDynamicParameters fspDynamicParam = DynamicParameters as GetChildDynamicParameters;
```
I thought I had a check for recurse at one time. It must have gotten lost in the shuffle. Fixed.

jeffbi commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1608,6 +1666,7 @@ protected override void InvokeDefaultAction(string path)
             DirectoryInfo directory,
             bool recurse,
             uint depth,
+            InodeTracker tracker,   // This will be non-null only if the user invoked the -FollowSymLinks switch parameter.
```
Fixed.

jeffbi commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1788,25 +1847,38 @@ protected override void InvokeDefaultAction(string path)
+                                else
                                 {
-                                    Dir(recursiveDirectory, recurse, depth - 1, nameOnly, returnContainers);
+                                    if (!tracker.TryVisitPath(recursiveDirectory.FullName))
```
Fixed.

jeffbi commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -7483,6 +7555,12 @@ internal sealed class GetChildDynamicParameters
 
         /// <summary>
+        /// Gets or sets the flag to follow symbolic links when recursing.
+        ///
```
Fixed.

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -7320,6 +7342,69 @@ private struct NetResource
+            /// </param>
+            /// <param name="visitIfNotAlready">
+            /// Flag to indicate whether the path should be marked as visited if
+            /// it has not already been visited.
```
Should we add the parameter if we don't use them?

jeffbi replied:
Normally I would say no. In this case the existence of the parameter and its name give a clue about the side effect of marking the path as visited.

daxian-dbw replied:
I give this some more thinking, and believes that the `TryVisitPath` signature is more natural in this case. But we need to follow the `Try<>` pattern.

Again, take `HashSet.Add(object)` as an example:
1. If the object is not in the set, add it and return true.
2. If the object is already in the set, return false which means it was not added in this call because it's already there.

Similarly, `InodeTracker.TryVisitPath(path)` should be
1. If the path is not visited yet, visit it and return true.
2. If the path is already visited, return false which means it was not visited in this call because it's already visited.

The code would be like:
```CSharp
internal bool TryVisitPath(string path)
{
    bool returnValue = false;
    if (InternalSymbolicLinkLinkCodeMethods.GetInodeData(path, out (UInt64, UInt64) inodeData))
    {
        returnValue = _visitations.Add(inodeData);
    }
    return returnValue;
}
```

@jeffbi @iSazonov what do you think?

iSazonov replied:
I like the TryVisitPath more the IsVisitPath.

jeffbi replied:
Works for me. I'll make the change.

daxian-dbw replied:
Awesome. Thank you both!

daxian-dbw commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1609,7 +1620,8 @@ protected override void InvokeDefaultAction(string path)
             bool nameOnly,
-            ReturnContainers returnContainers)
+            ReturnContainers returnContainers,
+            InodeTracker tracker)   // tracker will be non-null only if the user invoked the -FollowSymLinks switch parameter.
```
Minor comment: this comment is not accurate now -- `tracker` will be null when `-FollowSymlinks` is specified but `-Recurse` is not.

daxian-dbw reviewed (commented):
2 more comments

Commit:
Fix a comment

Commit:
Change InodeTracker's IsPathVisited to TryVisitPath

daxian-dbw reviewed (approved):
@jeffbi and @iSazonov Thank you both for the great work and thorough review!

iSazonov commented:
@jeffbi Many thanks for the great work!

jeffbi commented:
@iSazonov and @daxian-dbw Thanks for the review.

sba923 commented:
Very good job! 

P.S. I would probably have called the new parameter `-FollowSymlinks`
